drivers/sensor/bmi08x/bmi08x_accel_trigger.c

```diff
@@ -0,0 +1,153 @@
+LOG_MODULE_DECLARE(BMI08X_ACCEL, CONFIG_SENSOR_LOG_LEVEL);
+
+static void bmi08x_handle_drdy_acc(const struct device *dev)
+{
+	struct bmi08x_accel_data *bmi08x = dev->data;
+	struct sensor_trigger drdy_trigger = {
+		.type = SENSOR_TRIG_DATA_READY,
+	};
+
+#ifdef CONFIG_PM_DEVICE
+	enum pm_device_state state;
+
+	(void)pm_device_state_get(dev, &state);
+	if (state != PM_DEVICE_STATE_ACTIVE) {
+		return;
+	}
+#endif
+
+	if (bmi08x->handler_drdy_acc) {
+		bmi08x->handler_drdy_acc(dev, &drdy_trigger);
+	}
+}
+
+static void bmi08x_handle_interrupts_acc(void *arg)
+{
+	const struct device *dev = (device_ptr_t)arg;
+
+	bmi08x_handle_drdy_acc(dev);
+}
+
+#ifdef CONFIG_BMI08X_ACCEL_TRIGGER_OWN_THREAD
+static void bmi08x_acc_thread_main(void *arg1, void *unused1, void *unused2)
+{
+	k_thread_name_set(NULL, "bmi08x_acc_trig");
+
+	ARG_UNUSED(unused1);
+	ARG_UNUSED(unused2);
+
+	const struct device *dev = (device_ptr_t)arg1;
+	struct bmi08x_accel_data *bmi08x = dev->data;
+
+	while (1) {
+		k_sem_take(&bmi08x->sem, K_FOREVER);
+		bmi08x_handle_interrupts_acc((void *)dev);
+	}
+}
+#endif
+
+#ifdef CONFIG_BMI08X_ACCEL_TRIGGER_GLOBAL_THREAD
+static void bmi08x_acc_work_handler(struct k_work *work)
+{
+	struct bmi08x_accel_data *bmi08x = CONTAINER_OF(work, struct bmi08x_accel_data, work);
+
+	bmi08x_handle_interrupts_acc((void *)bmi08x->dev);
+}
+#endif
+
+static void bmi08x_acc_gpio_callback(const struct device *port, struct gpio_callback *cb,
+				     uint32_t pin)
+{
+	struct bmi08x_accel_data *bmi08x = CONTAINER_OF(cb, struct bmi08x_accel_data, gpio_cb);
+
+	ARG_UNUSED(port);
+	ARG_UNUSED(pin);
+
+#if defined(CONFIG_BMI08X_ACCEL_TRIGGER_OWN_THREAD)
+	k_sem_give(&bmi08x->sem);
+#elif defined(CONFIG_BMI08X_ACCEL_TRIGGER_GLOBAL_THREAD)
+	k_work_submit(&bmi08x->work);
+#endif
+}
+
+int bmi08x_trigger_set_acc(const struct device *dev, const struct sensor_trigger *trig,
+			   sensor_trigger_handler_t handler)
+{
+	struct bmi08x_accel_data *bmi08x = dev->data;
+
+	if ((trig->chan == SENSOR_CHAN_ACCEL_XYZ) && (trig->type == SENSOR_TRIG_DATA_READY)) {
+		bmi08x->handler_drdy_acc = handler;
+		return 0;
+	}
+
+	return -ENOTSUP;
+}
+
+int bmi08x_acc_trigger_mode_init(const struct device *dev)
+{
+	struct bmi08x_accel_data *bmi08x = dev->data;
+	const struct bmi08x_accel_config *cfg = dev->config;
+
+	if (!device_is_ready(cfg->int_gpio.port)) {
+		LOG_ERR("GPIO device not ready");
+		return -ENODEV;
+	}
+
+#if defined(CONFIG_BMI08X_ACCEL_TRIGGER_OWN_THREAD)
+	k_sem_init(&bmi08x->sem, 0, K_SEM_MAX_LIMIT);
+
+	k_thread_create(&bmi08x->thread, bmi08x->thread_stack,
+			CONFIG_BMI08X_ACCEL_THREAD_STACK_SIZE, bmi08x_acc_thread_main, (void *)dev,
+			NULL, NULL, K_PRIO_COOP(CONFIG_BMI08X_ACCEL_THREAD_PRIORITY), 0, K_NO_WAIT);
+#elif defined(CONFIG_BMI08X_ACCEL_TRIGGER_GLOBAL_THREAD)
+	bmi08x->work.handler = bmi08x_acc_work_handler;
+	bmi08x->dev = dev;
+#endif
+
+	/* set accel ints */
+	if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT1_MAP, cfg->int1_map) < 0) {
```

Comment:
I didn't find it in the datasheet of bmi088 and bmi085 BMI08X_REG_ACCEL_INT1_MAP: 0x56 and BMI08X_REG_ACCEL_INT2_MAP: 0x57. the datasheet said that 0x55-0x57 is reserved.

Reply:
it is there.... Just about every register that says "reserved" in that data actually does have a purpose (just not a very important one...) if you look in this Bosch public BMI08x bare metal driver.. you see those registers actually defined in there

Reply:
> 

I really don`t like "reserved" register, I can`t find those "reserved" register describe.

Reply:
it's here:
https://github.com/boschsensortec/BMI08x-Sensor-API/blob/631a966891885643f180e18e94870411b8f4c3fc/bmi08a.c#L2225

Reply:
> 

I think this is not data ready interrupt,  data ready interrupt setting is `set_accel_data_ready_int`
by writing BMI08X_REG_ACCEL_INT1_INT2_MAP_DATA register.

Reply:
I see... I'll replace it with this
```c
#if BMI08X_ACCEL_ANY_INST_HAS_DATA_SYNC
	if (config->data_sync != 0) {
		/* set accel ints */
		if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT1_MAP, cfg->int1_map) < 0) {
			LOG_ERR("Failed to map interrupts.");
			return -EIO;
		}
		if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT2_MAP, cfg->int2_map) < 0) {
			LOG_ERR("Failed to map interrupts.");
			return -EIO;
		}
	} else
#endif
	{
		uint8_t map_data = ((cfg->int2_map << BMI08X_ACCEL_INT2_DRDY_POS) |
			(cfg->int1_map << BMI08X_ACCEL_INT1_DRDY_POS));
		if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT1_INT2_MAP_DATA, map_data) < 0) {
			LOG_ERR("Failed to map interrupts.");
			return -EIO;
		}
	}
```

Reply:
I think it would be better to set the sensor synchronization function in the attribute. What do you think?
Macros can only be selected from one of two options, and using attribute functions can be set at any time.

> I see, I assume you're not using data sync?... I'll replace it with this
> 
> ```c
> #if BMI08X_ACCEL_ANY_INST_HAS_DATA_SYNC
> 	if (config->data_sync != 0) {
> 		/* set accel ints */
> 		if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT1_MAP, cfg->int1_map) < 0) {
> 			LOG_ERR("Failed to map interrupts.");
> 			return -EIO;
> 		}
> 		if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT2_MAP, cfg->int2_map) < 0) {
> 			LOG_ERR("Failed to map interrupts.");
> 			return -EIO;
> 		}
> 	} else
> #endif
> 	{
> 		uint8_t map_data = ((cfg->int2_map << BMI08X_ACCEL_INT2_DRDY_POS) |
> 			(cfg->int1_map << BMI08X_ACCEL_INT1_DRDY_POS));
> 		if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT1_INT2_MAP_DATA, map_data) < 0) {
> 			LOG_ERR("Failed to map interrupts.");
> 			return -EIO;
> 		}
> 	}
> ```

I think it would be better to set the sensor synchronization function in the attribute. What do you think?
Macros can only be selected from one of two options, and using attribute functions can be set at any time.

Reply:
Sorry, ...but I can't tell exactly what you're saying here??

Data sync is done in the DTS so it is instanced. You can have multiple BMI08X connected and one can have data sync on while the other doesn't. With a KConfig, it's either ALL or NONE.

Reply:
> ```c
> BMI08X_ACCEL_ANY_INST_HAS_DATA_SYNC
> ```

`bmi08x_attr_set`

Reply:
I emailed a bosch FAE about this a while ago... data sync can't be set a runtime